### PR TITLE
Change recheck API to accept recheck data as string

### LINKED ISSUE #383: Change recheck functions to accept Base64 encoding
Issue #332 is a feature request to only recheck the shrunken input.  To achieve this goal, PR #336 wants the recheck functions to also accept the path through the shrink tree to the shrunken input.  To make that change easier, we can modify the recheck functions to accept a Base64-encoded string of the recheck arguments.

[Here](https://stackoverflow.com/questions/11743160/how-do-i-encode-and-decode-a-base64-string) is how to encode/decode a Base64 string in C#.

This idea draws inspiration from [JetCheck](https://github.com/JetBrains/jetCheck), a property-based testing library in Java used by JetBrains on IntelliJ.

### PR DESCRIPTION
**Pull request by TysonMN**
Resolves #383

### CODE REVIEW

**Comment by TysonMN** [comment]
@Sweeney1309, you could also review this

**Comment by TysonMN** [comment]
The Fable build failed, probably because it doesn't have the Base64 functions I used.  Probably should just contact strings together for now.

File: src/Hedgehog/Report.fs [hunk 1]
```diff
@@ -29,6 +35,45 @@ type Report = {
...
+            |> Convert.FromBase64String
+            |> Encoding.UTF8.GetString
+
+    let private separator = "\n"
+
+    let serialize data =
+        data.Size.ToString ()         + separator
+        + data.Seed.Value.ToString () + separator
+        + data.Seed.Gamma.ToString ()
+        |> Base64.encode
```
Comment (dharmaturtle) [inline src/Hedgehog/Report.fs:61]:
```fsharp
[ string data.Size
  string data.Seed.Value
  string data.Seed.Gamma ]
|> String.concat separator
|> Base64.encode
```

Maybe?

File: src/Hedgehog/Report.fs [hunk 2]
```diff
@@ -9,10 +9,16 @@ type RecheckType =
     | None
     | CSharp
     | FSharp
-
-type FailureData = {
+    
+    
+[<Struct>]
+type RecheckData = internal {
```
Comment (dharmaturtle) [inline src/Hedgehog/Report.fs:15]: I'm not a big fan of the name RecheckData, from a consumer POV. Perhaps something like RecheckSeed? This doesn't need to be the name of the internal type... maybe just on the public API? e.g.

```fsharp
let recheck (recheckSeed: string) (p : Property<unit>) : unit =
```

Could perhaps make it even simpler:

```fsharp
let recheck (seed: string) (p : Property<unit>) : unit =
```

Comment (TysonMN) [reply src/Hedgehog/Report.fs:15]: I like `recheckSeed`.  I think `seed` is too confusing.

Comment (TysonMN) [reply src/Hedgehog/Report.fs:15]: But I think `recheckData` is better, even for the user.

File: src/Hedgehog/Report.fs (hunk 1 above)
Comment (TysonMN) [reply src/Hedgehog/Report.fs:62]: Sure.  Just include this change in a force push.

File: tests/Hedgehog.Tests/ReportTests.fs [hunk 3]
```diff
@@ -0,0 +1,22 @@
+﻿module Hedgehog.Tests.ReportTests
+
+open Hedgehog
+
+open TestDsl
+
+
+let reportTests = testList "Report tests" [
+    testCase "Roundtrip RecheckData serilzation" <| fun () ->
```
[the remark this answers is not in the document]
Comment (TysonMN) [reply tests/Hedgehog.Tests/ReportTests.fs]: Just include this change in a force push.

**Comment by TysonMN** [comment]
I removed the use of Base64 (which means that branch name isn't great any more, but whatever), and now the Fable tests pass.

**Comment by TysonMN** [comment]
And I change the separator to `_` so that failure message now looks like this
```
This failure can be reproduced by running:
> Property.recheck "0_7219121334209301472_6351764504400104151" <property>
```

File: src/Hedgehog/Report.fs [hunk 4]
```diff
@@ -29,6 +35,31 @@ type Report = {
...
+        try
+            let parts = s.Split([|separator|], StringSplitOptions.None)
+            let size = parts.[0] |> Int32.Parse
+            let seed =
+                { Value = parts.[1] |> UInt64.Parse
+                  Gamma = parts.[2] |> UInt64.Parse }
+            { Size = size
+              Seed = seed }
+        with e ->
+            raise (ArgumentException("Failed to deseralize RecheckData", e))
```
Comment (dharmaturtle) [inline src/Hedgehog/Report.fs:60]: `deserialize`

**Review by dharmaturtle — APPROVED** [review]
One last typo, but otherwise LGTM!

File: src/Hedgehog/Report.fs (hunk 4 above)
Comment (TysonMN) [reply src/Hedgehog/Report.fs]: Done

**Pull request merged**